lib/cli_common/cli_common/utils.py

```diff
@@ -45,3 +47,29 @@ def __exit__(self, *args):
                 future.cancel()
             raise e
         return super(ThreadPoolExecutorResult, self).__exit__(*args)
+
+
+class WatchdogException(Exception):
+    '''
+    Exception raised when a watchdog timer expires
+    '''
+
+
+@contextmanager
+def watchdog(max_runtime):
+    '''
+    Watchdog mechanism to limit execution time of a block of code
+    A WatchdogException is raised after the runtime is exceeded
+    '''
+    assert isinstance(max_runtime, int)
+    assert max_runtime > 0
+
+    def killer():
+        raise WatchdogException
+
+    timer = threading.Timer(max_runtime, killer)
```

Comment:
Won't this raise an exception in a separate thread, leaving the main thread still running?

Reply:
Maybe instead of doing this, you could do the clone in a separate thread, then join with a timeout:
```
t = Thread(target=clone)
t.start()
t.join(max_runtime)
if t.is_alive():
    # There was a timeout!
```

Reply:
I'm trying your solution, the `Timer` solution seems to work, but it may not be rock solid as i had an occurence of mercurial thread still running

Reply:
I think it works weel when using `sys.exit(1)` instead of the exception

Reply:
The thread solution might be better though, less of an hack.
Also I've tried:
```
import threading
import time
import sys


def killer():
    print('Ciao')
    sys.exit(1)

timer = threading.Timer(5, killer)
timer.start()

print('Start')
try:
    for i in range(15):
        time.sleep(1)
        print(i)
except:
    print('Caught')
print('End')
```

And it doesn't really stop the main thread:
```
Start
0
1
2
3
Ciao
4
5
6
7
8
9
10
11
12
13
14
End
```